druid-handler/src/java/org/apache/hadoop/hive/ql/io/DruidOutputFormat.java

```diff
@@ -0,0 +1,393 @@
+import io.druid.segment.realtime.appenderator.Appenderator;
+import io.druid.segment.realtime.appenderator.AppenderatorFactory;
+import io.druid.segment.realtime.appenderator.SegmentIdentifier;
+import io.druid.segment.realtime.appenderator.SegmentNotWritableException;
+import io.druid.segment.realtime.appenderator.SegmentsAndMetadata;
+import io.druid.segment.realtime.plumber.Committers;
+import io.druid.segment.realtime.plumber.CustomVersioningPolicy;
+import io.druid.storage.hdfs.HdfsDataSegmentPusher;
+import io.druid.storage.hdfs.HdfsDataSegmentPusherConfig;
+import io.druid.timeline.DataSegment;
+import io.druid.timeline.partition.LinearShardSpec;
+import org.apache.calcite.adapter.druid.DruidTable;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.conf.Constants;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.druid.DruidStorageHandlerUtils;
+import org.apache.hadoop.hive.druid.serde.DruidWritable;
+import org.apache.hadoop.hive.ql.exec.FileSinkOperator;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.RecordWriter;
+import org.apache.hadoop.mapred.Reporter;
+import org.apache.hadoop.util.Progressable;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+public class DruidOutputFormat<K, V> implements HiveOutputFormat<K, DruidWritable>
+{
+  private static final String DRUID_MAX_PARTITION_SIZE = "druid.maxPartitionSize";
+  private static final Integer DEFAULT_MAX_ROW_IN_MEMORY = 75000;
+  private final HiveConf hiveConf = SessionState.get().getConf();
+
+  public static class DruidRecordWriter implements RecordWriter<NullWritable, DruidWritable>,
+      org.apache.hadoop.hive.ql.exec.FileSinkOperator.RecordWriter
+  {
+    protected static final Logger log = LoggerFactory.getLogger(DruidRecordWriter.class);
+    private static final int DEFAULT_MAX_PARTITION_SIZE = 5000000;
+
+    private final Injector injector;
+    private final DataSchema dataSchema;
+    private final Appenderator appenderator;
+    private final RealtimeTuningConfig tuningConfig;
+    private final Map<Long, List<SegmentIdentifier>> segments = new HashMap<>();
+    private final Integer maxPartitionSize;
+
+    public DruidRecordWriter(
+        DataSchema dataSchema,
+        RealtimeTuningConfig realtimeTuningConfig,
+        Integer maxPartitionSize,
+        final Path segmentLocation
+    )
+    {
+      injector = Initialization.makeInjectorWithModules(
+          GuiceInjectors.makeStartupInjector(),
+          ImmutableList.<Module>of(new Module()
```

Comment:
IMO this is too large to embed within the method. Does it make sense to create a new class for DruidRecordWriter and then a new embedded class for this within it?